bpo-42195: Ensure consistency of Callable's __args__ in collections.abc and typing

Fidget-Spinner opened the pull request:
This patch fixes [bpo-42195](https://bugs.python.org/issue42195) via changing the `GenericAlias` builtin and subclassing for ``collections.abc.Callable``'s ``__class_getitem__``. The changes are:

1. Made `GenericAlias` subclassable.
2. Subclassed it in `collections.abc.Callable`'s `__class_getitem__` to allow for consistentency with `typing.Callable`'s `__args__`. This includes flattening the ``__args__``.


```python
>>> import collections.abc
>>> alias = collections.abc.Callable[[int, str], dict]

>>> alias
collections.abc.Callable[[int, str], dict]

>>> alias.__args__
(int, str, dict)
```
3. Added tests for the above.
4. Remove bad typing tests which subclassed from`` Callable[... T]``.
5. Loosened type checks for Callable's argtypes in ``Callable[[argtypes], resulttype]`` to prepare for PEP 612.

<!-- issue-number: [bpo-42195](https://bugs.python.org/issue42195) -->
https://bugs.python.org/issue42195
<!-- /issue-number -->

Fidget-Spinner committed:
Allow subclassing of GenericAlias, fix collections.abc.Callable's GenericAlias

Fidget-Spinner committed:
fix typing tests, add hash and eq methods

Fidget-Spinner committed:
Fix pickling

Fidget-Spinner committed:
whitespace

Fidget-Spinner committed:
Add test specifically for bpo

Fidget-Spinner committed:
update error message

Fidget-Spinner committed:
Appease test_site

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -20,6 +24,8 @@
            "Sequence", "MutableSequence",
            "ByteString",
+           # To allow for pickling, not actually meant to be imported.
+           "_CallableGenericAlias"
```
```suggestion
           "_CallableGenericAlias",
```

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +415,81 @@ def __subclasshook__(cls, C):
+    """ Internal class specifically for consistency between the ``__args__`` of
+    ``collections.abc.Callable``'s and ``typing.Callable``'s ``GenericAlias``.
+
+    See :issue:`42195`.
```
Let's not make a habit of referencing issues from comments/docstrings. It is almost always better to just explain the concern, if it isn't clear from reading the source.

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +415,81 @@ def __subclasshook__(cls, C):
+    def __new__(cls, *args, **kwargs):
+        if not isinstance(args, tuple) or len(args) != 2:
+            raise TypeError("Callable must be used as "
+                            "Callable[[arg, ...], result].")
```
Mention that it can also be `Callable[..., result]`?

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +415,81 @@ def __subclasshook__(cls, C):
+        _typ, _args = args
+        if not isinstance(_args, (tuple, EllipsisType)) or len(_args) != 2:
+            raise TypeError("Callable must be used as "
+                            "Callable[[arg, ...], result].")
```
Let's make sure the two error messages are not identical. Here I would call out that it can either be a list of types or an ellipsis.

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +415,81 @@ def __subclasshook__(cls, C):
+            if isinstance(arg, list):
+                ga_args.extend(arg)
+            else:
+                ga_args.append(arg)
```
Since you've already extracted `t_args` and `t_result`, why not do something like
```
if t_args == Ellipsis:
    ga_args = _args
else:
    ga_args = tuple(t_args) + (t_result,)
```
And you won't need the `tuple()` call below.

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +415,81 @@ def __subclasshook__(cls, C):
+    typically enough to uniquely identify a type.  For everything
+    else, we fall back on repr(obj).
+
+    Borrowed from :mod:`typing`.
```
Explain you can't import it from there because typing shouldn't be a dependency of this module. You don't need to copy the full docstring IMO.

gvanrossum reviewed (commented):
I suppose it would be even nicer if the subclass was written in C and used by builtins.callable as well? So we can write `callable[[int, int], str]`.

Fidget-Spinner committed:
Represent Callable __args__ via [tuple[args], result]

Fidget-Spinner committed:
add back tests for weakref, styling nits, add news

Fidget-Spinner committed:
remove redundant tuple checks leftover from old code

Fidget-Spinner committed:
Use _PosArgs instead of tuple

1.  allow equality between typing._PosArgs and collections.abc._PosArgs
2.  allow equality checks between genericalias subclasses
3.  add support to genericalias subclasses for union type expressions

Fidget-Spinner committed:
Fix typo and news

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -20,6 +24,11 @@
+           # meant to be imported.
+           "_CallableGenericAlias",
+           "_PosArgs",
+           "_PosArgsGenericAlias"
```
Ideally these would be imported from here into collections.abc but not exported by the latter, so that people writing `from collections.abc import *` won't see these -- pickle doesn't need them to be in `__all__`, but because of the way `collections/abc.py` is written it currently needs them to be in `__all__`.

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +418,94 @@ def __subclasshook__(cls, C):
+class _PosArgsGenericAlias(GenericAlias):
+    """ Internal class specifically to represent positional arguments in
+    ``_CallableGenericAlias``.
+    """
```
Since it's internal, maybe may this a comment instead of a docstring? (Also below.)

Then again, maybe it's not *entirely* internal, since the name is revealed by `repr()` and also (as you state) in pickles -- the former reveals it to users, while the latter makes it an eternal part of the API.

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +418,94 @@ def __subclasshook__(cls, C):
+    """
+    def __repr__(self):
+        return f"{__name__}._PosArgsGenericAlias" \
+               f"[{', '.join(_type_repr(t) for t in self.__args__)}]"
```
```suggestion
        arg = ', '.join(_type_repr(t) for t in self.__args__)
        return f"{__name__}._PosArgsGenericAlias[{arg}]"
```

However, I wonder if it shouldn't just return `f"PosArgs[{arg}]"`, since that is how you would create it.

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +418,94 @@ def __subclasshook__(cls, C):
+    def __eq__(self, other):
+        o_cls = other.__class__
+        if not (o_cls.__module__ == "typing" and o_cls.__name__
+                == "_GenericAlias" or isinstance(other, GenericAlias)):
```
```suggestion
        if not ((o_cls.__module__ == "typing"
                 and o_cls.__name__ == "_GenericAlias")
                or isinstance(other, GenericAlias)):
```
(Adding redundant parentheses to clarify the grouping of and/or, a common confusion.)

In addition, consider rearranging the logic to
```
if <condition> or <condition>:
    return <do the comparison>
return NotImplemented
```

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +418,94 @@ def __subclasshook__(cls, C):
+    def __class_getitem__(cls, item):
+        return _PosArgsGenericAlias(tuple, item)
+
+# _PosArgs = type("_PosArgs", (tuple, ), {})
```
Delete this?

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +418,94 @@ def __subclasshook__(cls, C):
+    """ Internal class specifically to represent positional arguments in
+    ``_CallableGenericAlias``.
+    """
+    def __repr__(self):
```
Probably add `__slots__ = ()`?

gvanrossum commented on Lib/test/test_genericalias.py:
```diff
@@ -302,6 +301,40 @@ def test_weakref(self):
+            self.assertEqual(alias.__args__, (_PosArgs[int, str], float))
+            self.assertEqual(alias.__parameters__, ())
+
+        with self.subTest("Testing nstance checks"):
```
```suggestion
        with self.subTest("Testing instance checks"):
```

gvanrossum commented on Lib/test/test_typing.py:
```diff
@@ -1818,7 +1818,7 @@ def __call__(self):
         ##     T2[int, str]
 
-        self.assertEqual(repr(C1[int]).split('.')[-1], 'C1[int]')
+        self.assertEqual(repr(C1[[int], int]).split('.')[-1], 'C1[[int], int]')
```
The point of this line was to check that `Callable[[T], T][int]` == `Callable[[int], int]`. Are you saying that `C1[int]` no longer works? That would violate the intended semantics of `__parameters__`. Same below.

Fidget-Spinner replied:
It still works, this is an accidental remnant from the ``__args__`` as ``((int, str), str)`` days :).

Fidget-Spinner replied:
Nevermind scratch that, it broke it all of it :(. The reason seems to be how `__parameters__` are passed for `types.GenericAlias`. ~Actually I don't understand how the typing/genericalias tests doesn't fail on main branch right now~. I also don't currently understand how typing does it for its generics (I'll take a look at it ).

```python
from typing import List, TypeVar
T = TypeVar('T')

class L1(List[T]): ...
print(L1.__parameters__) # (~T,)

class L1(list[T]): ...
print(L1.__parameters__) # AttributeError: type object 'L1' has no attribute '__parameters__'
```
Is this a bug or is the test at fault? I have a bad feeling about this...

EDIT:
From what I understand, typing's `__parameters__` is a class variable so generics in typing's `__class_getitem__` accesses its parents' `__parameters__` when computing it. I don't know if:
1. This is intended behavior in typing. (seems like it is).
2. The C implementation can do something like that. It looks like we need to change up a lot of code to add class properties.

Fidget-Spinner replied:
To be clear, the expression: ```Callable[[T], T][int] == Callable[[int], int]``` still works, and so does ``list[T][int] == list[int]``. Just that the test which subclasses it ``class C1(Callable[[T], T]): ...`` seems to fail.

gvanrossum commented on Lib/test/test_typing.py:
```diff
@@ -3058,7 +3058,7 @@ class C(Generic[T]): pass
         self.assertEqual(get_args(typing.Dict[int, Tuple[T, T]][Optional[int]]),
                          (int, Tuple[Optional[int], Optional[int]]))
-        self.assertEqual(get_args(Callable[[], T][int]), ([], int))
+        self.assertEqual(get_args(Callable[[], T][int]), (typing._PosArgs[()], int))
```
Hm, I think `get_args()` should expand the `PosArgs` away.

gvanrossum commented on Lib/typing.py:
```diff
@@ -114,6 +114,7 @@
     'Text',
     'TYPE_CHECKING',
     'TypeAlias',
+    '_PosArgs',  # Not meant to be imported, just for pickling.
```
I suspect that here it really is not needed in `__all__`.

gvanrossum commented on Lib/typing.py:
```diff
@@ -877,16 +878,19 @@ def __ror__(self, right):
+        if len(self.__args__) == 2 and t_args is Ellipsis:
             return super().__repr__()
+        t_args_repr = ('[]' if t_args.__args__ == ((),) else
+                       f'[{", ".join(_type_repr(a) for a in t_args.__args__)}]')
```
I would turn this into an if-statement, the expression form is not very readable.

gvanrossum commented on Lib/typing.py:
```diff
@@ -877,16 +878,19 @@ def __ror__(self, right):
     def __repr__(self):
         assert self._name == 'Callable'
-        if len(self.__args__) == 2 and self.__args__[0] is Ellipsis:
+        t_args = self.__args__[0]
```
If somehow one manages to get `Callable[()]` then this line would crash; the len() check was primarily meant to guard against that. (The *expected* length is always 2 of course, but it's possible that the rest of the implementation doesn't always guard against that.)

gvanrossum commented on Objects/genericaliasobject.c:
```diff
@@ -567,15 +567,41 @@ static PyGetSetDef ga_properties[] = {
 static PyObject *
 ga_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 {
+    gaobject *self;
```
I'd move this below to where it is initialized.

gvanrossum commented on Objects/genericaliasobject.c:
```diff
@@ -567,15 +567,41 @@ static PyGetSetDef ga_properties[] = {
     PyObject *arguments = PyTuple_GET_ITEM(args, 1);
-    return Py_GenericAlias(origin, arguments);
+
+    // almost the same as Py_GenericAlias' code, but to assign to self
```
I'd refactor this to avoid so much code duplication. Also, I'm sorry, I have forgotten this detail myself, I wonder what the difference is between what tp_alloc calls (PyType_GenericAlloc) and what Py_GenericAlias calls (PyObject_GC_New).

Fidget-Spinner replied:
Same here, from what I can see PyType_GenericAlloc calls _PyObject_GC_Malloc if Py_TPFLAGS_HAVE_GC flag is set, else it will call PyObject_MALLOC. While PyObject_GC_New calls _PyObject_GC_Malloc all the time. So it should be okay to replace one with another since that flag is set.

pablogsal replied:
Normally when calling `tp_alloc` directly instead of `PyObject_GC_New` is to allow subclasses to override the allocators, although using `tp_alloc` can be anecdotally slower.

gvanrossum commented on Misc/NEWS.d/next/Core and Builtins/2020-11-20-00-57-47.bpo-42195.HeqcpS.rst:
```diff
@@ -0,0 +1,7 @@
+and :class:`collections.abc.Callable` are now consistent.  Said ``__args__`` 
+are now in the form ``(_PosArgs[args], result)``.  For example, 
+``typing.Callable[[int], float].__args__`` now gives 
+``(typing._PosArgs[int], float)``.  Due to this change, 
```
This ignores the fact that typing._PosArg and collections.abc._PosArg are not the same object. (Not sure it's important, but the way it currently reads it can be misread to imply that.)

Also:
```suggestion
``(typing._PosArgs[int], float)``.  To allow to this change, 
```

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +418,94 @@ def __subclasshook__(cls, C):
+def _type_repr(obj):
+    """Return the repr() of an object, special-casing types (internal helper).
+
+    Borrowed from :mod:`typing` without importing since collections.abc
```
```suggestion
    Copied from :mod:`typing` since collections.abc
```

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +418,94 @@ def __subclasshook__(cls, C):
+    def __new__(cls, *args, **kwargs):
+        if not isinstance(args, tuple) or len(args) != 2:
+            raise TypeError("Callable must be used as Callable[[arg, ...], result]")
+        origin, _args = args
```
Because this is only called from `__class_getitem__` you can simplify this:
```suggestion
    def __new__(cls, origin, _args, **kwargs):
```
(The call from pickle also specifies two arguments, and if that ever were to fail it'd be better if it failed with an argument count error anyways.)

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +418,94 @@ def __subclasshook__(cls, C):
+        t_args, t_result = _args
+        if not isinstance(t_args, (list, EllipsisType)):
+            raise TypeError("Callable[args, result]: args must be a list "
+                            f"or Ellipsis. Got {_type_repr(t_args)}")
```
```suggestion
            raise TypeError(f"Callable[args, result]: args must be a list or Ellipsis. "
                            f"Got {_type_repr(t_args)}")
```

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +418,94 @@ def __subclasshook__(cls, C):
+                            f"or Ellipsis. Got {_type_repr(t_args)}")
+
+        ga_args = (_args if t_args is Ellipsis
+                   else (_PosArgs[tuple(t_args)], t_result))
```
Expand to if-statement? As soon as a conditional expression doesn't fit on a line it becomes awkward IMO. Also below.

gvanrossum reviewed (commented):
Sorry for leaving so many trivial code formatting comments. There are also a few important suggestions.

In general I am still not completely comfortable with this. In particular the change to how typing.Callable changes the form of its `__args__` makes me worried about backwards compatibility (that code has been the same for a long time, unlike collections.abc.Callable, which is new in 3.9.0).

I also worry that 3.9.1rc1 is already out, so the earliest we could get this in would be 3.9.2. But if we ever want to fix this I think 3.9.2 is still better than 3.10 (let alone never).

Something makes me think that we're probably better off keeping typing.Callable unchanged and changing collections.abc.Callable so that its `__args__` is a flat tuple. The trickery here is quite hard to follow. (Since I procrastinated on the review I feel this all the stronger. :-)

Fidget-Spinner committed:
Refactor C code to use less duplication

Fidget-Spinner committed:
Address most of Guido's reviews (tests failing on purpose)

Fidget-Spinner committed:
try to revert back to good old flat tuple __args__ days

Fidget-Spinner committed:
getting even closer

Fidget-Spinner committed:
finally done

Fidget-Spinner committed:
Update news

gvanrossum commented on Misc/NEWS.d/next/Core and Builtins/2020-11-20-00-57-47.bpo-42195.HeqcpS.rst:
```diff
@@ -1,7 +1,6 @@
 The ``__args__`` of the parameterized generics for :data:`typing.Callable`
 and :class:`collections.abc.Callable` are now consistent.  Said ``__args__`` 
```
```suggestion
and :class:`collections.abc.Callable` are now consistent.  The ``__args__`` 
```

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +413,80 @@ def __subclasshook__(cls, C):
 
+class _CallableGenericAlias(GenericAlias):
+    """ Internal class specifically for consistency between the ``__args__`` of
+    ``collections.abc.Callable``'s and ``typing.Callable``'s ``GenericAlias``.
```
I think this should not lead with the problem it solves; instead, lead with what the class is.
```suggestion
    """Represent `Callable[argtypes, resulttype]`.
    
    This sets ``__args__`` to the a tuple containing the flattening of ``argtypes``
    followed by ``resulttype``.

    Example: ``Callable[[int, str], float]`` sets ``__args__`` to ``(int, str, float`)`.
```
(Maybe also mention that ``Callable[..., str]`` produces ``(Ellipsis, int)``.)

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +413,80 @@ def __subclasshook__(cls, C):
+    def __new__(cls, origin, args):
+        try:
+            return cls.__getitem_type(origin, args)
+        except TypeError:
```
Is this *just* intended to catch the two explicit `raise` statements in `__getitem_type()` below? Then why spend any time in that function producing a nice error message? Why not just write `return GenericAlias(origin, args)` there?

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +413,80 @@ def __subclasshook__(cls, C):
+
+    def __reduce__(self):
+        args = self.__args__
+        if not (len(args) == 2 and args[0] is ...):
```
```suggestion
        if not (len(args) == 2 and args[0] is Ellipsis):
```

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +413,80 @@ def __subclasshook__(cls, C):
+            # class C1(typing.Callable[[T], T]): ...
+            # C1[int]
+            #
+            # Otherwise that will raise a TypeError.
```
More typical examples are just `Callable[int]` and `Callable[int, int]`. (I had to think about why you chose this particular example (which isn't even legal according to PEP 484). Honestly do we need this try/except at all?

Fidget-Spinner replied:
Honestly, I wanted to remove the entire try except block and yes the example is admittedly contrived. The problem is with the try except removed, the following test in test_typing fails:
```python
    def test_extended_generic_rules_subclassing(self):
        class T1(Tuple[T, KT]): ...
        class T2(Tuple[T, ...]): ...
        class C1(Callable[[T], T]): ...
        class C2(Callable[..., int]):
            def __call__(self):
                return None

        [...]

        self.assertEqual(repr(C1[int]).split('.')[-1], 'C1[int]')
        [...]
```
As you can see, that typing test doesn't conform to PEP484 either. Should I instead change the test?

gvanrossum replied:
> Honestly, I wanted to remove the entire try except block and yes the example is admittedly contrived. The problem is with the try except removed, the following test in test_typing fails:
> 
> ```python
>     def test_extended_generic_rules_subclassing(self):
>         class T1(Tuple[T, KT]): ...
>         class T2(Tuple[T, ...]): ...
>         class C1(Callable[[T], T]): ...
>         class C2(Callable[..., int]):
>             def __call__(self):
>                 return None
> 
>         [...]
> 
>         self.assertEqual(repr(C1[int]).split('.')[-1], 'C1[int]')
>         [...]
> ```
> 
> As you can see, that typing test doesn't conform to PEP484 either. Should I instead change the test?

Oh, that's unfortunate. Yes, let's delete those two classes and their uses from that test.

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +413,80 @@ def __subclasshook__(cls, C):
+        if obj.__module__ == 'builtins':
+            return obj.__qualname__
+        return f'{obj.__module__}.{obj.__qualname__}'
+    if obj is ...:
```
```suggestion
    if obj is Ellipsis:
```

gvanrossum commented on Lib/test/test_genericalias.py:
```diff
@@ -302,6 +301,33 @@ def test_weakref(self):
                 alias = t[int]
                 self.assertEqual(ref(alias)(), alias)
 
+    def test_abc_callable(self):
```
Can you also add a test that e.g. `F = Callable[[int, T], T]` followed by `F[str]` indeed returns `Callable[[int, str], str]`?

gvanrossum commented on Objects/genericaliasobject.c:
```diff
@@ -564,9 +564,50 @@ static PyGetSetDef ga_properties[] = {
     {0}
 };
 
+// Helper to create inheritable or non-inheritable gaobjects
```
I don't understand this comment.

gvanrossum reviewed (commented):
I'm sorry, I haven't fully reviewed the C code yet, but I figured you'd like to hear from me before the weekend. :-)

Fidget-Spinner committed:
Address review partially

* Add tests for callable typevar substitution
* Update docstrings and news

Co-Authored-By: Guido van Rossum <gvanrossum@users.noreply.github.com>

Fidget-Spinner committed:
Address review fully, update news and tests, remove try-except block

Fidget-Spinner committed:
Borrowed references don't need decref

Fidget-Spinner committed:
improve _PyArg_NoKwnames error handling, add union and subclass tests

Fidget-Spinner committed:
Don't change getargs, use _PyArg_NoKeywords instead

Fidget-Spinner committed:
Merge remote-tracking branch 'upstream/master' into abc-callable-ga

Fidget-Spinner committed:
remove stray whitespace

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -422,29 +422,10 @@ class _CallableGenericAlias(GenericAlias):
 
-    @classmethod
-    def __getitem_type(cls, origin, args):
+    def __new__(cls, origin, args):
```
Okay, for 3.10 I think this is the right solution.

For 3.9, I think we need to issue a `DeprecationWarning`(using the warnings module) with the same message (for both `TypeError`s below) and return what was returned in 3.9.0.

gvanrossum commented on Lib/test/test_typing.py:
```diff
@@ -1807,10 +1807,9 @@ def barfoo2(x: CT): ...
-                return None
+        class C1(typing.Container[T]):
+            def __contains__(self, item):
+                return False
```
So in the 3.9 backport this file should be untouched.

gvanrossum reviewed (commented):
Very close for the Python code. For the C code I am still not sure that this is the best refactoring. Basically we have some shared code (if not tuple, tuplify), then some non-shared code (allocate object), then some more shared code (set attributes), then some non-shared code (gc-track).

I think I may have led you astray by asking to share the code between the two. Or maybe twe could change the order to

1. allocate
2. tuplify
3. set attrs
4. gc-track

and then only put (2) and (3) in the shared helper function.

What do you think?

Fidget-Spinner committed:
refactor C code, add deprecation warning for 3.9

gvanrossum commented on Lib/_collections_abc.py:
```diff
@@ -409,6 +413,80 @@ def __subclasshook__(cls, C):
+                              DeprecationWarning)
+                return GenericAlias(origin, args)
+            else:
+                raise exc
```
This should probably be just `raise` (not `raise exc`). But really, I think we just need to check in different code in the 3.9 branch (and we can refine the 3.9 code after the 3.10 code is landed). Sorry about the misunderstanding.

Fidget-Spinner replied:
Ok, and no worries, I misinterpreted that. Btw for the sake of easier backporting any future changes (I foresee ParamSpec being one of them), I'm going to separate out the method call.

Fidget-Spinner replied:
Woops I just remembered ParamSpec is 3.10, I meant any bugfixes to Callable 😉 .

gvanrossum commented on Objects/genericaliasobject.c:
```diff
@@ -564,9 +564,36 @@ static PyGetSetDef ga_properties[] = {
 static PyObject *
 ga_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 {
+    assert(type != NULL && type->tp_alloc != NULL);
```
Do we still need this check?

Fidget-Spinner replied:
I'm not too sure, I did it because dict_new in dictobject.c does it.

gvanrossum commented on Objects/genericaliasobject.c:
```diff
@@ -564,9 +564,36 @@ static PyGetSetDef ga_properties[] = {
+setup_ga(gaobject *alias, PyObject *origin, PyObject *args) {
+    if (alias == NULL) {
+        return 0;
+    }
```
We don't need this check, both call sites already check for that.

Fidget-Spinner committed:
remove redundant check in C code, and try except in __new__

Fidget-Spinner committed:
remove check

Fidget-Spinner committed:
Loosen type checks for Callable args, cast to PyObject in genericaliasobject.c

Fidget-Spinner committed:
update news to mention about removing validation in argtypes

gvanrossum commented on Lib/test/test_typing.py:
```diff
@@ -446,14 +446,17 @@ def test_cannot_instantiate(self):
+        # with self.assertRaises(TypeError):
+        #     Callable[[()], int]
+        # with self.assertRaises(TypeError):
+        #     Callable[[int, 1], 2]
```
Let's not leave commented-out code around. People can read the git logs.

gvanrossum commented on Lib/test/test_typing.py:
```diff
@@ -1861,10 +1862,11 @@ def test_type_erasure_special(self):
+        ## class MyCall(Callable[..., T]):
+        ##     def __call__(self): return None
+        ## self.assertIs(MyCall[T]().__class__, MyCall)
+        ## self.assertEqual(MyCall[T]().__orig_class__, MyCall[T])
```
I would drop this as well.

gvanrossum reviewed (commented):
I think it's ready except for the commented-out blocks of code.

Also, I'd like to ask @pablogsal to see if he has time to review your C code (he has often caught things I missed in C code!).

Fidget-Spinner committed:
remove commented out code

gvanrossum reviewed (approved):
Actually @pablogsal may be busy, let's see if @brandtbucher is interested.

pablogsal reviewed (approved):
I have reviewed the C code and it looks good to me! Good job @Fidget-Spinner !
